server/src/main/java/org/elasticsearch/action/support/DefaultShardOperationFailedException.java

```diff
@@ -105,7 +106,7 @@ protected XContentBuilder innerToXContent(XContentBuilder builder, Params params
         builder.field("status", status.name());
         if (reason != null) {
             builder.startObject("reason");
-            ElasticsearchException.generateThrowableXContent(builder, params, cause);
+            ElasticsearchException.generateThrowableXContent(builder, params, cause, Sets.newHashSet(this));
```

Comment:
It's a bit unfortunate that this one requires us to add a public interface exposing `Set<Throwable> seenExceptions`. As mentioned above, an alternative approach that isolates the change a bit more might be to track observed instances in the builder itself.